test/support/metrics_case.ex

```diff
@@ -0,0 +1,108 @@
+defmodule RequestCache.Support.MetricsCase do
+  @moduledoc """
+  This module defines the setup for tests requiring metrics tests.
+
+  Available tags:
+    - `telemetry_events`: list the Telemetry events to listen
+    - `metrics`: Specify the list of Telemetry.Metrics to used (format: [Module, :function, [args]])
+
+  Available assertions:
+    - `assert_telemetry_event(name, measurements, metadata \\ %{})`
+    - `assert_metric(name, measurement, metadata \\ %{})`
+
+  ## Example:
+
+  @tag telemetry_events: [[:user, :subscription, :email_confirmation]],
+       metrics: [RequestCache.Support.Metrics, :metrics, []]
+  test "my test" do
+    ...
+    assert_telemetry_event([:user, :subscription, :email_confirmation], %{count: 1}, %{result: :error})
+    assert_metric([:user, :subscription, :email_confirmation, :count], 1, %{success: false})
+  end
+
+  Sauce: https://elixirforum.com/t/testing-and-telemetry-events-how-to-test-if-they-are-sent/28273/4
+  """
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      import RequestCache.Support.MetricsCase
+    end
+  end
+
+  setup tags do
+    if telemetry_events = tags[:telemetry_events] do
+      metrics = get_metrics_from_tag(tags)
+
+      self = self()
+
+      groups = Enum.group_by(metrics, & &1.event_name)
+
+      :telemetry.attach_many(
+        tags[:test],
+        telemetry_events,
+        fn name, measurements, metadata, _config ->
+          send(self, {:telemetry_event, name, measurements, metadata})
+
+          # Send related metrics
+          if Enum.count(metrics) > 0 do
+            Enum.each(Map.get(groups, name, []), fn metric ->
+              send(
```

Comment:
Wowza this got super nested, 3 levels of nesting max! Make use of helper functions 😉

Reply:
Oi, yuck. Fixed.